clients/pkg/logentry/stages/limit.go

```diff
@@ -0,0 +1,202 @@
+package stages
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"regexp"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/time/rate"
+)
+
+const (
+	ErrLimitStageEmptyConfig        = "limit stage config must contain at least one of `source`, `expression`, `older_than` or `longer_than`"
+	ErrLimitStageInvalidConfig      = "limit stage config error, `value` and `expression` cannot both be defined at the same time."
+	ErrLimitStageInvalidRegex       = "limit stage regex compilation error: %v"
+	ErrLimitStageInvalidRateOrBurst = "limit stage failed to parse rate or burst"
+)
+
+var ratelimitDropReason = "ratelimit_drop_stage"
+
+type LimitConfig struct {
+	Source     *string `mapstructure:"source"`
+	Rate       float64 `mapstructure:"rate"`
+	Burst      int     `mapstructure:"burst"`
+	Drop       bool    `mapstructure:"drop"`
+	Value      *string `mapstructure:"value"`
+	Expression *string `mapstructure:"expression"`
+	regex      *regexp.Regexp
+}
```

Comment:
Sorry, I should have spotted this early. Would you mind adding documentation in (https://github.com/grafana/loki/blob/main/docs/sources/clients/promtail/pipelines.md#stages) as well? I'm not sure I understand `rate` and `burst`.